### PR TITLE
If ignore label exist don`t add namespace

### PR DESCRIPTION
**Pull request by mudit-01**
PR partially fixes https://github.com/openservicemesh/osm/issues/4445

Signed-off-by: mudit singh <mudit.singh@india.nec.com>

**Description**:

**Testing done**: Yes

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| CI System                  | [ ] |
| CLI Tool                   | [X] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Demo                       | [ ] |
| Documentation              | [ ] |
| Egress                     | [ ] |
| Ingress                    | [ ] |
| Install                    | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| Performance                | [ ] |
| SMI Policy                 | [ ] |
| Security                   | [ ] |
| Sidecar Injection          | [ ] |
| Tests                      | [ ] |
| Upgrade                    | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

2. Is this a breaking change?

3. Has documentation corresponding to this change been updated in the [osm-docs](https://github.com/openservicemesh/osm-docs) repo (if applicable)?

### CODE REVIEW

**Comment by mudit-01** [comment]
The e2e_ignore_namespace_test.go test is failing as it is trying to add an ignored namespace to the mesh that is forbidden by this PR. 
My query is can we remove this test? After this PR adding an ignored namespace to a mesh will not be possible, hence disabling sidecar injection will not be required.
What are your views?

**Comment by allenlsy** [comment]
@mudit-01 I think what the original e2e test wants to verify is still valid, which is sidecar injection should not work for ignored namespace. But the way it sets the environment is not right. It:

* creates a namespace with ignore label on it.
* adds it to the mesh

Rather than doing so, it should:

* create a namespace
* use `osm ignore` to ignore the namespace

Then proceed with adding a pod to the namespace and check the existence of sidecar.

Can you update the e2e test case in this PR also?

**Comment by shashankram** [comment]
> @mudit-01 I think what the original e2e test wants to verify is still valid, which is sidecar injection should not work for ignored namespace. But the way it sets the environment is not right. It:
> 
> * creates a namespace with ignore label on it.
> * adds it to the mesh
> 
> Rather than doing so, it should:
> 
> * create a namespace
> * use `osm ignore` to ignore the namespace
> 
> Then proceed with adding a pod to the namespace and check the existence of sidecar.
> 
> Can you update the e2e test case in this PR also?

@allenlsy there is nothing wrong with how the current e2e test sets the label directly on the namespace. The logic was meant to test the ignore label functionality in relation to sidecar injection, and not the CLI behavior. We do not need to test CLI behavior in e2e as a unit test should suffice to test CLI.

**Comment by mudit-01** [comment]
@shashankram Should I update the e2e test to `Expect(Td.AddNsToMesh(true, ignoreNs)).NotTo(Succeed())` as once the ignore lable is found it should not add the namespace to the mesh?

**Comment by johnsonshi** [comment]
Summarizing what @jaellio and I discussed:

* `osm namespace ignore` currently patches a `openservicemesh.io/ignore: true` label to a namespace when run.
* Mutating Webhook and Validating Webhook use the *existence* of the `openservicemesh.io/ignore` label (not the boolean value) as the top-level admission criteria. If the `ignore` label does not exist, the webhooks further check criteria by looking at the value of `openservicemesh.io/enable-sidecar-injection`.
* From running `osm namespace ignore --help`, the intent of the command is to (1) allow a namespace to be added to osm but (2) prevent further NEW sidecars to be injected.
```
$ osm namespace ignore --help
This command will prevent a namespace or a set of namespaces from
participating in the mesh. Automatic sidecar injection on pods
belonging to the given namespace or set of namespaces will be prevented.
The command will not remove previously injected sidecars on pods belonging
to the given namespaces.
```
* The `enable-sidecar-injection` label and `ignore` labels currently seem to do duplicate work. One should supersede the other, and one should be deprecated imho.

**Comment by shashankram** [comment]
> Proposal:
> 
> * `osm namespace ignore [namespace]` should patch a `openservicemesh.io/enable-sidecar-injection: false` label to a namespace instead of an `openservicemesh.io/ignore: true` label.

The `ignore` label is working the way it is intended to. It is meant to ignore resources (mutating webhooks, ingresses, etc.). We should not be tweaking the sidecar injection label in relation to the ignore label. The ignore label should supersede the sidecar injection annotation.

**Comment by mudit-01** [comment]
@allenlsy @shashankram @johnsonshi any suggestions on this or this is ok to merge?

**Review by shashankram — APPROVED** [review]

**Review by jaellio — APPROVED** [review]

File: cmd/cli/namespace_add.go [hunk 1]
```diff
@@ -121,6 +123,11 @@ func (a *namespaceAddCmd) run() error {
 			continue
 		}
 
+		// if ignore label exits don`t add namespace
+		if namespace.ObjectMeta.Labels[constants.IgnoreLabel] == trueValue {
```
Comment (shashankram) [inline cmd/cli/namespace_add.go:127]:
```suggestion
		if val, ok := namespace.ObjectMeta.Labels[constants.IgnoreLabel]; ok && val == trueValue {
```

**Pull request merged**